spec/index.md

```diff
@@ -5033,6 +5094,179 @@ S with
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'take_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+New_snapshot = Snapshot {
+  wasm_state = canisters[A.canister_id].wasm_state;
+  raw_module = canisters[A.canister_id].raw_module;
+  wasm_module = canisters[A.canister_id].wasm_module;
+  chunk_store = canisters[A.canister_id].chunk_store;
+}
+Cycles_reserved = cycles_to_reserve(S, A.canister_id, S.compute_allocation[A.canister_id], S.memory_allocation[A.canister_id], New_snapshot, S.canisters[A.canister_id])
+New_balance = S.balances[A.canister_id] - Cycles_used - Cycles_reserved
+New_reserved_balance = S.reserved_balances[A.canister_id] + Cycles_reserved
+New_reserved_balance ≤ S.reserved_balance_limits[A.canister_id]
+
+liquid_balance(
+  New_balance,
+  New_reserved_balance,
+  freezing_limit(
+    S.compute_allocation[A.canister_id],
+    S.memory_allocation[A.canister_id],
+    S.freezing_threshold[A.canister_id],
+    memory_usage_wasm_state(New_state) +
+      memory_usage_raw_module(A.wasm_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(New_snapshot),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+```
+
+State after  
+
+```html
+
+S with
+    snapshots[A.canister_id] = New_snapshot
+    balances[A.canister_id] = New_balance
+    reserved_balances[A.canister_id] = New_reserved_balance
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin;
+        response = Reply (candid(SnapshotId));
+        refunded_cycles = M.transferred_cycles;
+      }
+
+```
+
+
+#### IC Management Canister: Load snapshots
+
+
+Only the controllers of the given canister can load a snapshot.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'load_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+New_balance = S.balances[A.canister_id] - Cycles_used
+liquid_balance(
+  New_balance,
+  S.reserved_balances[A.canister_id],
+  freezing_limit(
+    S.compute_allocation[A.canister_id],
+    S.memory_allocation[A.canister_id],
+    S.freezing_threshold[A.canister_id],
+    memory_usage_wasm_state(S.canisters[A.canister_id]) +
+      memory_usage_raw_module(A.wasm_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(S.snapshots[A.canister_id]),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+```
+
+State after  
+
+```html
+
+S with
+    canisters[A.canister_id] = {
+      wasm_state = S.snapshots[A.canister_id].wasm_state;
+      raw_module = S.snapshots[A.canister_id].raw_module;
+      module = S.snapshots[A.canister_id].wasm_module;
```

Comment:
since we're already parsing the custom section, we could parse the wasm module from the raw module and drop wasm_module from `Snapshot = {`: module = parse_wasm_mod(S.snapshots[A.canister_id].raw_module)